// @tsed/common/lib/platform-builder/middlewares/ContextMiddleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const platform_1 = require("../../platform");
const onFinished = require("on-finished");
const uuidv4 = require("uuid/v4");
const defaultReqIdBuilder = () => uuidv4().replace(/-/gi, "");
class ContextMiddleware {
    constructor(injector) {
        this.injector = injector;
        const { level, maxStackSize, ignoreUrlPatterns = [], reqIdBuilder = defaultReqIdBuilder } = injector.settings.logger || {};
        this.level = level;
        this.maxStackSize = maxStackSize;
        this.ignoreUrlPatterns = ignoreUrlPatterns;
        this.reqIdBuilder = reqIdBuilder;
    }
    static onClose(err, response) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { req: request } = response;
            yield request.ctx.emit("$onResponse", request, response);
            yield request.ctx.destroy();
            delete request.ctx;
            delete request.log;
        });
    }
    use(request, response, next) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { level, ignoreUrlPatterns, maxStackSize } = this;
            const id = this.reqIdBuilder();
            request.ctx = new platform_1.RequestContext({
                id,
                logger: this.injector.logger,
                url: request.originalUrl || request.url,
                ignoreUrlPatterns,
                level,
                maxStackSize,
                injector: this.injector
            });
            // deprecated
            request.log = request.ctx.logger;
            onFinished(response, ContextMiddleware.onClose);
            yield this.injector.emit("$onRequest", request, response);
            next();
        });
    }
}
exports.ContextMiddleware = ContextMiddleware;
