// @tsed/common/lib/mvc/pipes/ValidationPipe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const di_1 = require("@tsed/di");
const RequiredValidationError_1 = require("../errors/RequiredValidationError");
const ValidationService_1 = require("../services/ValidationService");
let ValidationPipe = class ValidationPipe {
    constructor(validationService) {
        this.validationService = validationService;
    }
    transform(value, metadata) {
        this.checkIsRequired(value, metadata);
        if (this.shouldValidate(metadata)) {
            const { collectionType } = metadata;
            const type = metadata.type || metadata.collectionType;
            this.validationService.validate(value, type, collectionType);
        }
        return value;
    }
    shouldValidate(param) {
        return !!(param.type || param.collectionType);
    }
    checkIsRequired(value, metadata) {
        if (metadata.isRequired(value)) {
            throw RequiredValidationError_1.RequiredValidationError.from(metadata);
        }
        return true;
    }
};
ValidationPipe = tslib_1.__decorate([
    di_1.Injectable({
        type: "validator"
    }),
    tslib_1.__metadata("design:paramtypes", [ValidationService_1.ValidationService])
], ValidationPipe);
exports.ValidationPipe = ValidationPipe;
