// @tsed/exceptions/lib/core/Exception.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Exception extends Error {
    constructor(status = 500, message = "", origin) {
        super(message);
        /**
         * Exception base name
         * @type {string}
         */
        this.name = "HTTP_EXCEPTION";
        /**
         * Exception type
         * @type {string}
         */
        this.type = "HTTP_EXCEPTION";
        /**
         * Headers must be sent to the response
         */
        this.headers = {};
        this.status = status;
        this.message = message;
        this.setOrigin(origin);
    }
    setHeaders(headers) {
        this.headers = Object.assign(Object.assign({}, this.headers), headers);
        return this;
    }
    setHeader(key, value) {
        this.headers[key] = value;
        return this;
    }
    setOrigin(origin) {
        if (origin) {
            if (origin instanceof Error) {
                this.origin = origin;
                this.message = `${this.message}, innerException: ${this.origin.message}`.trim();
            }
            else if (typeof origin === "string") {
                this.origin = new Error(origin);
                this.message = `${this.message}, innerException: ${this.origin.message}`.trim();
            }
            else {
                this.body = origin;
            }
        }
    }
    toString() {
        return (this.name + "(" + this.status + "): " + this.message + " ").trim();
    }
}
exports.Exception = Exception;
class HTTPException extends Exception {
}
exports.HTTPException = HTTPException;
