// @tsed/common/lib/platform/services/RouteService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const di_1 = require("@tsed/di");
const Platform_1 = require("./Platform");
/**
 * `RouteService` is used to provide all routes collected by annotation `@Controller`.
 * @deprecated Use Platform instead
 */
let RouteService = class RouteService {
    constructor(platform) {
        this.platform = platform;
    }
    get routes() {
        return this.platform.routes;
    }
    addRoutes(routes) {
        return this.platform.addRoutes(routes);
    }
    /**
     * Add a new route in the route registry
     * @param endpoint
     * @param token
     */
    addRoute(endpoint, token) {
        return this.platform.addRoute(endpoint, token);
    }
    /**
     * Get all routes built by TsExpressDecorators and mounted on Express application.
     * @returns {IRouteDetails[]}
     */
    getRoutes() {
        return this.platform.getRoutes();
    }
    /**
     * @deprecated Use getRoutes instead of
     */
    getAll() {
        return this.getRoutes();
    }
};
RouteService = tslib_1.__decorate([
    di_1.Injectable(),
    tslib_1.__metadata("design:paramtypes", [Platform_1.Platform])
], RouteService);
exports.RouteService = RouteService;
